routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,376 @@
+          MYTHROW(RoutingException,
+                  ("Dublicated joint id should be less then count, convertedId =", convertedId,
+                   ", count =", m_count, ", originalId =", originalId));
+        }
+        writer.Write(1, 1);
+        Write32(writer, m_count - convertedId);
+        return;
+      }
+
+      m_convertedIds[originalId] = m_count;
+      writer.Write(0, 1);
+      ++m_count;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+
+  private:
+    Joint::Id m_count = 0;
+    unordered_map<Joint::Id, Joint::Id> m_convertedIds;
+  };
+
+  class JointIdDecoder final
+  {
+  public:
+    JointIdDecoder(Joint::Id begin) : m_count(begin) {}
+    template <class Source>
+    Joint::Id Read(BitReader<Source> & reader)
+    {
+      uint8_t const repeatBit = reader.Read(1);
+      if (repeatBit)
+      {
+        uint32_t const delta = Read32(reader);
+        if (delta > m_count)
+          MYTHROW(RoutingException, ("Joint id delta", delta, "> count =", m_count));
+
+        return m_count - delta;
+      }
+
+      return m_count++;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+  private:
+    Joint::Id m_count;
+  };
+
+  class JointIdConverter final
+  {
+  public:
+    JointIdConverter(Joint::Id numJoints) { m_ids.assign(numJoints, Joint::kInvalidId); }
+    Joint::Id Convert(Joint::Id jointIdInFile)
+    {
+      if (jointIdInFile >= m_ids.size())
+        MYTHROW(RoutingException,
+                ("Saved joint id", jointIdInFile, " out of bounds", m_ids.size()));
+
+      Joint::Id & idForMemory = m_ids[jointIdInFile];
+      if (idForMemory == Joint::kInvalidId)
+      {
+        idForMemory = m_count;
+        ++m_count;
+      }
+
+      return idForMemory;
+    }
+
+    Joint::Id GetCount() const { return m_count; }
+  private:
+    Joint::Id m_count = 0;
+    vector<Joint::Id> m_ids;
+  };
+
+  class SectionSerializer final
+  {
+  public:
+    explicit SectionSerializer(VehicleMask mask) : m_mask(mask) {}
+
+    uint64_t GetBufferSize() const { return m_buffer.size(); }
+    VehicleMask GetMask() const { return m_mask; }
+    uint32_t GetNumRoads() const { return m_featureIds.size(); }
+
+    void AddRoad(uint32_t featureId) { m_featureIds.push_back(featureId); }
+    void SortRoads() { sort(m_featureIds.begin(), m_featureIds.end()); }
+    void SerializeToBuffer(IndexGraph const & graph,
+                           unordered_map<uint32_t, VehicleMask> const & masks,
+                           JointIdEncoder & jointEncoder);
+
+    template <class Sink>
+    void Flush(Sink & sink)
+    {
+      sink.Write(m_buffer.data(), m_buffer.size());
+      m_buffer.clear();
+    }
+
+  private:
+    VehicleMask const m_mask;
+    vector<uint32_t> m_featureIds;
+    vector<uint8_t> m_buffer;
+  };
+
+  template <typename Sink>
+  static void Write32(BitWriter<Sink> & writer, uint32_t value)
+  {
+    bool const success = coding::GammaCoder::Encode(writer, static_cast<uint64_t>(value));
```

Comment:
Gamma coder returns false only when asked to encode zero. So, you may put an assertion here that value is not zero, and get rid of exception.